static/posts/pattern-matching/post.md

```diff
@@ -0,0 +1,952 @@
+  } else if (shape.type == ...) {
+    return ...
+  } else {
+    printf("Missing handler for shape with type: %d\n", shape.type);
+    exit(1);
+  }
+}
+
+// ...
+```
+
+One way to improve the readability slightly is to just use `if` statements:
+
+```c
+// ...
+
+double shape_area(Shape shape) {
+  Dimensions ds = shape.dimensions;
+  if (shape.type == CIRCLE) {
+    return M_PI * ds.r * ds.r;
+  }
+  if (shape.type == RECTANGLE) {
+    return ds.w * ds.h;
+  }
+  if (shape.type == TRIANGLE) {
+    double s = (ds.a + ds.b + ds.c) / 2;
+    return sqrt(s * (s - ds.a) * (s - ds.b) * (s - ds.c));
+  }
+  if (shape.type == ...) {
+    return ...;
+  }
+  if (shape.type == ...) {
+    return ...;
+  }
+  if (shape.type == ...) {
+    return ...;
+  }
+  if (shape.type == ...) {
+    return ...;
+  }
+  if (shape.type == ...) {
+    return ...;
+  }
+  if (shape.type == ...) {
+    return ...;
+  }
+  printf("Missing handler for shape with type: %d\n", shape.type);
+  exit(1);
+}
+
+// ...
+```
+
+Another option is to use `switch`/`case` which may still be a bit easier to
+read:
+
+```c
+// ...
+
+double shape_area(Shape shape) {
+  Dimensions ds = shape.dimensions;
+  switch (shape.type) {
+    case CIRCLE:
+      return M_PI * ds.r * ds.r;
+    case RECTANGLE:
+      return ds.w * ds.h;
+    case TRIANGLE:
+      double s = (ds.a + ds.b + ds.c) / 2;
+      return sqrt(s * (s - ds.a) * (s - ds.b) * (s - ds.c));
+    case ...:
+      return ...;
+    default:
+      printf("Missing handler for shape with type: %d\n", shape.type);
+      exit(1);
+  }
+}
+
+// ...
+```
+
+However, I'd argue that it doesn't solve the main problem we're dealing with,
+and it gives you another way to shoot yourself in the foot: If you or someone
+else forgets to return or break in a case, then the calculations may be
+incorrect again, leading to another difficult-to-debug problem. So now, in each
+place where you may be processing shapes like this, you have to remember to add
+handlers when new shapes are added and always remember to return or break.
+
+These problems aren't just limited to C.
+[Languages that descend from C](https://en.wikipedia.org/wiki/List_of_C-family_programming_languages)
+tend to inherit its syntax and imperative programming style, including
+`if`/`else` and `switch`/`case`.
+
+## First-class Pattern Matching
+
+As stated earlier, many modern programming languages have support for pattern
+matching. The checks we're looking for can be done at compile time rather than
+at runtime. You should check if your favorite programming language has support
+for pattern matching so that you can avoid the pitfalls discussed above.
+
+### Haskell
+
+Let's start with Haskell, one of the most influential and "pure" functional
+programming languages.
+
+Pattern matching is how you define functions in Haskell. I believe this is one
+of the main reasons why people struggle with understanding Haskell, especially
+people who have only used imperative programming.
```

Comment:
change this "people" to "those" - since you've already used "people" once in this sentence. Or you could use "programmers" or another specific computer programming title